### ISSUE TITLE
Support for Mockito

### ISSUE THREAD

**Issue by tom91136** [dodie/scott#25]
Currently, Mockito fails while stubbing as scott tries to call toString while mockito is capturing method calls. 
This could be partially resolved with a check in `hu.advancedweb.scott.runtime.track.StateRegistry#objectToString`:

```java
	//...
	} else if (MockUtil.isMock(value)) {
		// While setting up mocks with when(...), invoking any method on the mock will
		// cause it to be recorded thus breaking the when() statement.
		return constructMockitoMockObjectString(value);
	}else if (MockUtil.isSpy(value)){
		return "Mockito Spy:" + MockUtil.getMockName(value);
	}	//...

	public static String constructMockitoMockObjectString(Object value){
		return "Mockito Mock("+ MockUtil.getMockSettings(value).getTypeToMock()+")";
	}
```
Along with the required dependency:

```xml
    <dependency>
        <groupId>org.mockito</groupId>
        <artifactId>mockito-core</artifactId>
        <version>2.2.28</version>
    </dependency>
```

But it still doesn't solve the case where method calls are inlined when stubbing.

**Comment by dodie** [comment]
Nice find, thank you! I've tried the fix you provided and it indeed works well. However, there is
one problem with this solution: adding Mockito as a dependency (even if Scott ships with a
specific version) might result in compatibility problems with projects using other Mockito versions.

I think rather than statically depending on ```MockUtils```, we should detect mocks and spies more dynamically
to work around this issue. For example we could check via the reflection API if MockUtils available at run-time
with the required methods, and if it is then call it. 

For the inlined method calls problem, could you please provide an example for it when it goes wrong?

**Comment by tom91136** [comment]
Sure, `Class.forName` seems standard so we can use that. I'll prepare a pull if you want.

For the inline method problem, consider the following test:

```java

    import static org.mockito.Mockito.doAnswer;
    import static org.mockito.Mockito.mock;
    import static org.mockito.Mockito.times;
    import static org.mockito.Mockito.verify;
    import static org.mockito.Mockito.when;

    interface Foo {
    
    	String bar();
    
    	void baz(Consumer<String> consumer);
    
    }
    
    @Test
    public void simple_mock() throws Exception {
    	Foo foo = mock(Foo.class);
    	when(foo.bar()).thenReturn("42");
    	assertEquals(foo.bar(), "42");
    	verify(foo, times(1)).bar();
    }
    
    class Holder<T> {
    
    	final T t;
    
    	Holder(T t) {
    		this.t = t;
    	}
    
    	@Override
    	public String toString() {
    		return t != null ? t.toString() : "(null)";
    	}
    }
    
    @Test
    public void field_mock() throws Exception {
    	Foo foo = mock(Foo.class);
    	Holder<Foo> holder = new Holder<>(foo);
    	when(holder.t.bar()).thenReturn("42");
    	assertEquals(holder.t.bar(), "42");
    	verify(holder.t, times(1)).bar();
    }
```

Test `simple_mock` passed fine but the `field_mock` failed with the following:

    java.lang.Throwable: 
      48|      @Test
      49|      public void field_mock() throws Exception {
      50|          Foo foo = mock(Foo.class);  // foo=Mockito Mock(interface hu.advancedweb.scott.MockitoTest$Foo)
      51|          Holder<Foo> holder = new Holder<>(foo);  // holder=Mock for Foo, hashCode: 1668016508
      52|*         when(holder.t.bar()).thenReturn("42");  // UnfinishedStubbingException: Unfinished stubbing detected here:
        |                                                  // -> at hu.advancedweb.scott.MockitoTest.field_mock(MockitoTest.java:52)
        |                                                  // 
        |                                                  // E.g. thenReturn() may be missing.
        |                                                  // Examples of correct stubbing:
        |                                                  //     when(mock.isOk()).thenReturn(true);
        |                                                  //     when(mock.isOk()).thenThrow(exception);
        |                                                  //     doThrow(exception).when(mock).someVoidMethod();
        |                                                  // Hints:
        |                                                  //  1. missing thenReturn()
        |                                                  //  2. you are trying to stub a final method, which is not supported
        |                                                  //  3: you are stubbing the behaviour of another mock inside before 'thenReturn' instruction if completed
      53|          assertEquals(holder.t.bar(), "42");
      54|          verify(holder.t, times(1)).bar();
      55|      }

The main issue here is that scott is trying to call `toString` on things that will be used in `when()`.
I'm still not 100% sure on the exact cause but I think `when()` is basically a singleton stored in a `ThreadLocal`. When scott tries to call `toString` on objects that have been registered with `when()` the actual call sequence gets modified into something like this (pseudocode):

    52|*         when(
                           holder.t.toString()
                           holder.t.bar()
                            ).thenReturn("42");

This is hard to fix as scott has to basically know that some field will be used for stubbing in the future.
I'll still investigating this so I'll get back with more details ASAP.

**Comment by dodie** [comment]
I've played with the Mockito tests, and I found that calling toString on a mock (even indirectly) is only invalid during stubbing.  For illustration, execute the following tests without Scott:

```java
	@Test
	public void simple_mock_ok() throws Exception {
		Foo foo = mock(Foo.class);
		System.out.println(foo.toString()); // Print mock before stubbing is OK
		OngoingStubbing<String> stubbing = when(foo.bar());
		stubbing.thenReturn("42");
		System.out.println(foo.toString()); // Print mock after stubbing is OK
		assertEquals(foo.bar(), "42");
		verify(foo, times(1)).bar();
	}

	@Test
	public void simple_mock_failing() throws Exception {
		Foo foo = mock(Foo.class);
		OngoingStubbing<String> stubbing = when(foo.bar());
		System.out.println(foo.toString()); // Print mock during stubbing is NOT OK
		stubbing.thenReturn("42");
		assertEquals(foo.bar(), "42");
		verify(foo, times(1)).bar();
	}
```

Limiting Scott's toString calls seems necessary, but I think this problem can't be solved in the ```StateRegistry```, because it's not important what object is being called with ```toString```, but the location of the call. Scott should not inject toString calls investigate variable values when the test calls Mockito related methods.

The ```hu.advancedweb.scott.instrumentation.transformation.StateEmitterTestMethodVisitor``` class is responsible for augmenting the test classes with tracking code The ```visitMethodInsn``` is called for every method invocation, this method calls the ```instrumentToTrackVariableState``` method to track every variable and field. I think the filtering could be done based on the ```owner``` parameter.

If you are interested in this and have time pull requests are appreciated.

**Comment by tom91136** [comment]
Thanks for the details, I'm interested and will submit a pull request as soon as I complete support for Mockito along with tests and all that.

**Comment by dodie** [comment]
I've been thinking about this issue lately, and I was wondering that maybe we could solve it by making the recording part more resilient, by swallowing any exception in the ```StateRegistry#objectToString``` and simply not tracking anything in those cases.

I think this would solve the current issue with Mockito without having to bake Mockito related edge-case handling into the code, and would probably mitigate issues with other libraries/code that does nasty things with ```toString```.

What do you think, would this fix be sufficient for this issue?

**Comment by tom91136** [comment]
Sorry for the long delay. I had work on exams and studies I'm still a student. As I am also University of Bristol's teaching assistance for an undergraduate course, I was planning on adding this to one of the student courseworks so that they can better test their code without using a debugger. The cut off time for that particular coursework was around the end of January so unfortunately I was not able to integrate scott into the coursework due to issues with Mockito. The next coursework that I can reintroduce scott will start development at the start of June. 

Personally, I think Mockito is a rather common testing dependency so I think supporting Mockito explicitly would be a better idea. I'll see what I can do in June.

**Comment by dodie** [comment]
I meant no pressure, I just thought that I check in, to see if I can help with something. The news about you considering Scott for your previous course is really great, I am glad to hear that!

Also, I had some time on my hands, so I investigated the issue a bit further. I added some code to fix the current problems with Mockito, and I'd like to share my findings.

I started with your solution by adding an isMock check to the ```StateRegistry#objectToString```  method. This indeed fixed the first case you've provided, but I could not figure out how could it be enhanced to fix the second one. The problem is, that Mockito uses a ThreadLocal variable to keep track of the methods called to a mock, and uses this information when you configure mocked methods. Also, the signature of the ```org.mockito.Mockito.when(T methodCall)``` is a bit misleading, as it doesn't do much with the supplied object.

For example, this piece of code:

```
Foo foo = mock(Foo.class);
when(foo.bar()).thenReturn("42");
assertEquals("42", foo.bar());
``` 

is more or less equal with this one:

```
Foo foo = mock(Foo.class);
foo.bar();
when(null).thenReturn("42");
assertEquals("42", foo.bar());
```

Scott adds an extra toString call after every method call, so it looks something like this (pseudo-code):
```
Foo foo = mock(Foo.class);
track(foo.toString);
foo.bar();
track(foo.toString);
OngoingStubbig s = when(null);
track(foo.toString);
s.thenReturn("42");
track(foo.toString);
assertEquals("42", foo.bar());
track(foo.toString);
```
(There are some extra tracking for each variable assignment that I omitted.)

The second ```toString``` call breaks the mock, as, from that point, Mockito thinks that we tried to mock the toString method on ```foo```.

I modified two things to solve this:

- Added a check to the [instrumentation code](https://github.com/dodie/scott/commit/77ddaa4e7d1ed8ad0cb50ea7c6ba6558e03b6f0a#diff-71220351e79d14a1945cd26ce4b7054eR108) responsible for tracking method calls, to leave methods provided by Mockito (```when```, ```mock```, ```thenReturn```...) alone.
- [Switched the order](https://github.com/dodie/scott/commit/77ddaa4e7d1ed8ad0cb50ea7c6ba6558e03b6f0a#diff-71220351e79d14a1945cd26ce4b7054eR126) of the original method call, and the tracker code. So the extra toString method call will precede the original call, so we don't confuse Mockito. (This modification involved some modification with the tracking of the line numbers as well.)

After the modifications, the mock get's successfully configured in both the holder and the simple scenario, and the code of the test looks like the following:

```
Foo foo = mock(Foo.class);
track(foo.toString);
foo.bar();
when(null).thenReturn("42");
track(foo.toString);
assertEquals("42", foo.bar());
track(foo.toString);
```

At that point, there was a problem with ```verify```, as the following snippet

```
verify(foo, times(1)).bar();
```

now looks like the following with Scott:
```
verify(foo, times(1))
track(foo.toString)
foo.bar();
```

So Mockito now thinks that we are trying to verify ```toString```, and then ```bar```, and that's not working because verify()-ing toString [results in an exception](https://github.com/mockito/mockito/wiki/FAQ).
So finally, [I've added an exception handling logic](https://github.com/dodie/scott/commit/77ddaa4e7d1ed8ad0cb50ea7c6ba6558e03b6f0a#diff-777fdbaaf8928ae25993d1dd9d343fc7R206) to the ```StateRegistry``` that ignores MockitoExceptins.

For now, it seems that Scott works correctly with Mockito, at least for the use-cases I've tried.
I also added some [tests for it](https://github.com/dodie/scott/commit/77ddaa4e7d1ed8ad0cb50ea7c6ba6558e03b6f0a#diff-b08b0800751cc5cd90dd61acd3d5e1ccR1).

If you have time, could you verify that Scott works correctly for your use case?

**Comment by dodie** [comment]
I've released the 3.0.0 version from Scott that supports Mockito.
If you have any further issues with the Mockito integration, feel free to open an issue.
